### PR TITLE
feat: safe Nat indexed Array functions using get_elem_tactic

### PR DESCRIPTION
**Pull request by Shreyas4991**
Adds safe Nat indexed variants of Array functions: `getN`, `setN`, `swapN`, `swapAtN`, and `eraseIdxN`.

### CODE REVIEW

**Comment by Shreyas4991** [comment]
awaiting-review

File: Batteries/Data/Array/Basic.lean [hunk 1]
```diff
@@ -153,6 +153,58 @@ Unsafe implementation of `attachWith`, taking advantage of the fact that the rep
...
+should prove the index bound.
+-/
+
+/--
+`getN a i` gets the `iᵗʰ` element of `a` if valid.
+By default, the proof of validity `h : i < a.size` is supplied by the `get_elem_tactic`.
+If this fails, a proof must be supplied at the function call.
+-/
+abbrev getN (a : Array α) (i : Nat) (h : i < a.size := by get_elem_tactic) : α :=
+  Array.get a ⟨i, h⟩
```
Comment (dupuisf) [inline Batteries/Data/Array/Basic.lean:170]: For this particular one, is there a scenario where one would use `as.getN i` instead of just `as[i]`?

Comment (Shreyas4991) [reply Batteries/Data/Array/Basic.lean:170]: It's to be sure that `h` is being used and not some variable potentially captured from the environment (variable declarations).

Comment (dupuisf) [reply Batteries/Data/Array/Basic.lean:170]: I don't understand: if `h` is generated by `get_elem_tactic` then it's the same thing as `as[i]`, and if it's provided explicitly, then it's the same as `as[i]'h`, no?

Comment (Shreyas4991) [reply Batteries/Data/Array/Basic.lean:170]: How does `a[i]` generate a proof of validity?

Comment (Shreyas4991) [reply Batteries/Data/Array/Basic.lean:170]: Sorry edited that one and lost the old message. But yeah crashes happen in both `as[i]` and `as.getN i` for invalid indices. Only `as.get` seems safe.

Comment (Shreyas4991) [reply Batteries/Data/Array/Basic.lean:170]: Should this be deleted?

Comment (fgdorais) [reply Batteries/Data/Array/Basic.lean:170]: I think so. Sometimes it makes sense to have a non-notation-based version for special cases but this doesn't strike me as one.

Comment (Shreyas4991) [reply Batteries/Data/Array/Basic.lean]: Done.

Comment (fgdorais) [reply Batteries/Data/Array/Basic.lean]: Same idea applies to #793, I think.

Comment (Shreyas4991) [reply Batteries/Data/Array/Basic.lean]: Done.

**Review by dupuisf — APPROVED** [review]

**Review by fgdorais — APPROVED** [review]

**Pull request merged**